src/main/java/cn/nukkit/Player.java

```diff
@@ -2767,6 +2768,14 @@ public void onCompletion(Server server) {
                             this.dataPacket(entityEventPacket);
                             Server.broadcastPacket(this.getViewers().values(), entityEventPacket);
                             break;
+                        case EntityEventPacket.ENCHANT:
+                            if (entityEventPacket.eid != this.id) {
+                                break;
+                            }
+
+                            int levels = entityEventPacket.data; // Sent as negative number of levels lost
+                            this.setExperience(this.exp, this.expLevel + levels);
```

Comment:
You should check if the value received isn't higher than 0 at least

Reply:
What if a modified client does not it? I think it would be better to check the amount of lapis used in the enchantment transactions and remove XP depending on that.

Reply:
If I remember correctly player sends this before the transaction so it's possible to check

Reply:
I will have to double check the order of packets, but I'm not sure if we'd know the lapis cost based on how it's currently set up. Crafting and Enchanting need overhaul in 2.0 

I will take a closer look into packet order tomorrow

Reply:
Added a sanity check to make sure the data is less than 0, should be okay for 1.0 for now. In 2.0 we need to rework how the InventoryTransaction packets are processed.